documentation/source/protocol_extension.rst

```diff
@@ -0,0 +1,91 @@
+How to write a protocol extension
+=================================
+
+Cocotb take care of the core communication between the simulator and testbench.
+To use some known busses like wishbone, avalon, spi, i2c,... An extension is required.
+
+This is a guideline that will explain how to write a python module protocol
+extension for cocotb.
+
+Naming convention
+-----------------
+
+A cocotb extension module must begin by `cocotb_` and terminate with protocol
+name.
+
+For exemple, with spi the module extension is `cocotb_spi`.
+
+Bus driver
+----------
+
+A Bus driver is a python class that factorize common behavior and check for
+presence of some signals.
+
+A Bus driver is an active class that be driven in the main testbench. Example is
+given with avalon bus in `cocotb/drivers/avalon.py`.
+
+With AvalonMaster class we can generate stimulis for reading and writing on bus.
+
+The main class is named `BusDriver` and is defined in `cocotb/drivers/__init__.py`
+
+TODO: give an example.
+
+Bus Monitor
+-----------
+
+A Bus monitor is mainly a passive class that will monitor the bus.
+
+TODO: explain monitor
+
```

Comment:
I feel here is a good place to add the following information based on https://github.com/cocotb/cocotb/pull/1190#discussion_r347848900

```
Handling Errors
---------------

It may not be clear when to raise an exception and when to use ``log.error()``.

Use ``raise`` if the caller called your function in an invalid way,
and it doesn't make sense to continue.

Use ``log.error()`` if the hardware itself is misbehaving,
and throwing an error immediately would leave it an invalid state.

Even if you do eventually throw an exception (if you weren't able to do it immediately),
you should also ``log.error()`` so that the simulation time of when things went wrong
is recorded.

TL;DR: ``log.error()`` is for humans only, raise is for both humans and code.
```

Reply:
Added